extension/extension.go

```diff
@@ -16,6 +16,11 @@ import (
 // to the service, examples: health check endpoint, z-pages, etc.
 type Extension = component.Component
 
+// TODO this can even be DependentComponent, but not sure if there's a use case for that.
+type DependentExtension interface {
+	Dependencies() []component.ID
```

Comment:
As I mentioned in https://github.com/open-telemetry/opentelemetry-collector/pull/8733#issuecomment-1783637186, this capability is only applicable for the Jager-specific use case. It doesn't cover the way components depend on extensions in the Collector core/contrib repos.

Reply:
Sorry but I don't understand how you see the duck-typing solution work when you have more than one dependency implementing the same interface.

With this PR you can use this config:
```
extensions:
  complex_extension_1:
    storage: filestorage
  complex_extension_2:
    storage: dbstorage
  filestorage:
    ...
  dbstorage:
    ...
```
The "complex" extensions here will implement DependentExtension and return an ID of the other extension they depend on, which they can read from their user-provided config.

Reply:
> The "complex" extensions here will implement DependentExtension and return an ID of the other extension they depend on, which they can read from their user-provided config.

After this change, the `complex` extensions have to know about every implementation of the storage extension interface: `filestorage` and `dbstorage`. If I have another storage extension in my own distribution of OTel collector, the `complex` won't be able to work with it.

Reply:
Another simple option is to introduce some kind of priority class for the extensions that would be considered in the order of initialization

Reply:
> After this change, the complex extensions have to know about every implementation of the storage extension interface: filestorage and dbstorage. If I have another storage extension in my own distribution of OTel collector, the complex won't be able to work with it.

On the contrary, it does not need to know anything about implementations at all, only the interface they implement. The actual IDs of the dependent extensions are provide by the user in the configuration.

Reply:
I see what you mean. I missed the fact that user config is already available at this point. Sorry for the confusion. This approach looks good to me.